Sources/PenguinTesting/StdlibProtocolTests.swift

```diff
@@ -164,37 +164,161 @@ extension Comparable {
+fileprivate final class CollectionChecker<C, E>
+  : SequenceChecker<C, E> where SequenceCheckConstraints<C,E>: Any {}
+
+extension CollectionChecker: SequenceRefinementChecker where C : Collection {
+  /// Runs the XCTest.
+  func checkSemantics() {
+    subject.checkCollectionSemantics(expecting: exampleContents)
+  }
+}
+
+/// A checker for sequences also declared to conform to `MutableCollection`.
+fileprivate final class MutableCollectionChecker<Subject, NewContents>
+  : SequenceChecker<Subject, NewContents>
+  where SequenceCheckConstraints<Subject, NewContents>: Any {}
+
+extension MutableCollectionChecker: SequenceRefinementChecker where Subject : MutableCollection {
+  convenience init(_ subject: Subject, writing newContents: NewContents) {
+    self.init(subject, expecting: newContents)
+  }
+  
+  /// Runs the XCTest.
+  func checkSemantics() {
+    subject.checkMutableCollectionSemantics(writing: exampleContents)
+  }
+}
+
+/// A checker for sequences also declared to conform to `BidirectionalCollection`.
+fileprivate final class BidirectionalCollectionChecker<C,E>
+  : SequenceChecker<C, E> where SequenceCheckConstraints<C,E>: Any {}
+
+extension BidirectionalCollectionChecker: SequenceRefinementChecker
+  where C : BidirectionalCollection
+{
+  /// Runs the XCTest.
+  func checkSemantics() {
+    subject.checkBidirectionalCollectionSemantics(expecting: exampleContents)
+  }
+}
+
+/// A checker for sequences also declared to conform to `RandomAccessCollection`.
+fileprivate final class RandomAccessCollectionChecker<C, E>
+  : SequenceChecker<C, E> where SequenceCheckConstraints<C,E>: Any {}
 
-/// A “generic type predicate” that detects whether a type is a
-/// `BidirectionalCollection`.
-fileprivate struct IsBidirectionalCollection<C> {}
-extension IsBidirectionalCollection: True where C : BidirectionalCollection {  }
+extension RandomAccessCollectionChecker: SequenceRefinementChecker
+  where C : RandomAccessCollection
+{
+  /// Runs the XCTest.
+  func checkSemantics() {
+    subject.checkRandomAccessCollectionSemantics(expecting: exampleContents)
+  }
+}
 
-/// A “generic type predicate” that detects whether a type is a
-/// `RandomAccessCollection`.
-fileprivate struct IsRandomAccessCollection<C> {}
-extension IsRandomAccessCollection: True where C : RandomAccessCollection {  }
+extension Sequence where Element: Equatable {
+  ///  XCTests that `self` has proper semantics for the known refinements of `Sequence` to which
+  ///  `Self` has been declared to conform.
+  ///
+  /// - Note: use this method from generic testing contexts where you may not know the declared
+  ///   conformances of the concrete `Sequence` type under test.  When you *do* know the declared
+  ///   conformances, it is best to (additionally) use `isCollection`, `isBidirectional`, et al., to
+  ///   check that they are as you expect.
+  ///
+  /// - Requires: if `Self: MutableCollection`, `expectedContents` is not a palindrome.
```

Comment:
Another requirement is "if `Self: Collection`, `self.count >= 2`".

Note that there are count=1 vectors, so I want to remove this requirement (and the palindrome requirement). But that's orthogonal to this PR so I could do it myself later.

Reply:
A collection with `count == 1` is always a palindrome.

Reply:
Yes, it's always a palindrome so it's impossible to test a `count == 1` mutable collection, so I'll need to do something to make it work for vectors with a single scalar.

Also, my "another requirement" comment is still unaddressed.

Reply:
Oh, I see, you're right. I thought I was addressing it by saying that, but it doesn't cover the non-mutablecollection case.

Reply:
https://github.com/saeta/penguin/commit/a968027f80b997fdf29968afab746a6b5f5e444d